api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.db.hibernate;
+
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.hibernate.Query;
+import org.hibernate.SessionFactory;
+import org.openmrs.Diagnosis;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+
+/**
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
+ * @since 2.2
+ */
+public class HibernateDiagnosisDAO implements DiagnosisDAO {
+	
+	private static final Log log = LogFactory.getLog(DiagnosisDAO.class);
+
+	/**
+	 * Hibernate session factory
+	 */
+	private SessionFactory sessionFactory;
+
+	/**
+	 * Set session factory
+	 *
+	 * @param sessionFactory
+	 */
+	public void setSessionFactory(SessionFactory sessionFactory) {
+		this.sessionFactory = sessionFactory;
+	}
+
+	/**
+	 * Saves the diagnosis.
+	 *
+	 * @param diagnosis the diagnosis to save.
+	 * @return the saved diagnosis.
+	 */
+	@Override
+	@Transactional
+	public Diagnosis saveDiagnosis(Diagnosis diagnosis) {
+		sessionFactory.getCurrentSession().saveOrUpdate(diagnosis);
+		return diagnosis;
+	}
+
+	/**
+	 * Gets all active diagnoses related to the specified patient.
+	 *
+	 * @param patient the patient whose active diagnoses are being queried.
+	 * @return all active diagnoses associated with the specified patient.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getActiveDiagnoses(Patient patient) {
+		Query query = sessionFactory.getCurrentSession().createQuery(
+			"from Diagnosis d where d.patient.patientId = :patientId and d.voided = false "
```

Comment:
I'd imagine HQL should figure out that it needs to compare the database ids so d.patient = :patient would work

Reply:
What are active Diagnoses? I'd think getDiagnoses(Patient) by default returns non voided one

Reply:
My current implementation of getDiagnoses gets all diagnoses for a given encounter without consideration of whether they are voided or not and that is why there is the method to getActiveDiagnoses which returns non voided ones only